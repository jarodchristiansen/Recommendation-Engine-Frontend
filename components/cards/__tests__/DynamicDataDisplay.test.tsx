// import "@testing-library/jest-dom";
// import {
//   render,
//   screen,
//   fireEvent,
//   waitFor,
//   act,
// } from "@testing-library/react";
// import DynamicDataDisplay from "@/components/cards/DynamicDataDisplay";

// // Mock fetch call
// global.fetch = jest.fn(() =>
//   Promise.resolve({
//     json: () =>
//       Promise.resolve({
//         items: [
//           {
//             id: "1",
//             name: "Song One",
//             artist_name: "Artist One",
//             image_url: "/image1.jpg",
//             album: {
//               images: [{ url: "/image1.jpg" }],
//             },
//           },
//           {
//             id: "2",
//             name: "Song Two",
//             artist_name: "Artist Two",
//             image_url: "/image2.jpg",
//             album: {
//               images: [{ url: "/image2.jpg" }],
//             },
//           },
//         ],
//       }),
//   })
// ) as jest.Mock;

// describe("DynamicDataDisplay Component", () => {
//   const defaultProps = {
//     endpoint: "/api/recommendations",
//     type: "track",
//     selectedSongs: [],
//     onSelectSong: jest.fn(),
//     onClearSelection: jest.fn(),
//   };

//   beforeAll(() => {
//     global.ResizeObserver = class {
//       observe() {}
//       unobserve() {}
//       disconnect() {}
//     };
//   });

//   it("displays loading animation while fetching data", () => {
//     render(<DynamicDataDisplay {...defaultProps} />);
//     expect(
//       screen.getByText("Hold tight, we're crunching some numbers")
//     ).toBeInTheDocument();
//     expect(
//       screen.getByText(
//         "This may take up to 60 seconds, but it’ll be worth the wait!"
//       )
//     ).toBeInTheDocument();
//   });

//   it("fetches and renders CardGrid when data is loaded and type is 'track'", async () => {
//     await act(async () => {
//       render(<DynamicDataDisplay {...defaultProps} />);
//     });

//     await waitFor(() => {
//       expect(screen.getByText("Song One")).toBeInTheDocument();
//     });

//     expect(screen.getByText("Song Two")).toBeInTheDocument();
//   });

//   it("renders RecommendCardGrid when type is 'recommendations'", async () => {
//     const props = { ...defaultProps, type: "recommendations" };

//     await act(async () => {
//       render(<DynamicDataDisplay {...props} />);
//     });

//     await waitFor(() => {
//       expect(screen.getByText("Artist One")).toBeInTheDocument();
//     });
//   });

//   it("handles error state", async () => {
//     (global.fetch as jest.Mock).mockImplementationOnce(() =>
//       Promise.resolve({
//         json: () => Promise.resolve({ error: "Failed to fetch data" }),
//       })
//     );

//     await act(async () => {
//       render(<DynamicDataDisplay {...defaultProps} />);
//     });

//     await waitFor(() => {
//       expect(screen.getByText("Failed to fetch data")).toBeInTheDocument();
//     });
//   });

//   it("handles item click and selection", async () => {
//     const mockOnSelectSong = jest.fn();

//     await act(async () => {
//       render(
//         <DynamicDataDisplay
//           {...defaultProps}
//           onSelectSong={mockOnSelectSong}
//           selectedSongs={[]}
//         />
//       );
//     });

//     await waitFor(() => {
//       expect(screen.getByText("Song One")).toBeInTheDocument();
//     });

//     // Simulate clicking on a song
//     const songOne = screen.getByText("Song One").closest("div");
//     fireEvent.click(songOne!);

//     expect(mockOnSelectSong).toHaveBeenCalled();
//   });

//   it("displays 'Clear Selection' button when there are selected songs", async () => {
//     const selectedSongs = [
//       {
//         id: "1",
//         name: "Song One",
//         artist_name: "Artist One",
//         image_url: "/image1.jpg",
//       },
//     ];

//     await act(async () => {
//       render(
//         <DynamicDataDisplay {...defaultProps} selectedSongs={selectedSongs} />
//       );
//     });

//     expect(screen.getByText("Clear Selection")).toBeInTheDocument();

//     // Simulate clearing selection
//     const clearButton = screen.getByText("Clear Selection");
//     fireEvent.click(clearButton);

//     expect(defaultProps.onClearSelection).toHaveBeenCalled();
//   });

//   it("renders correctly when no data is returned", async () => {
//     (global.fetch as jest.Mock).mockImplementationOnce(() =>
//       Promise.resolve({
//         json: () => Promise.resolve({ items: [] }),
//       })
//     );

//     await act(async () => {
//       render(<DynamicDataDisplay {...defaultProps} />);
//     });

//     await waitFor(() => {
//       expect(screen.queryByText("Song One")).not.toBeInTheDocument();
//       expect(screen.queryByText("Song Two")).not.toBeInTheDocument();
//     });
//   });
// });

import "@testing-library/jest-dom";
import {
  render,
  screen,
  fireEvent,
  waitFor,
  act,
} from "@testing-library/react";
import DynamicDataDisplay from "@/components/cards/DynamicDataDisplay";

// Mock fetch call
global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () =>
      Promise.resolve({
        items: [
          {
            id: "1",
            name: "Song One",
            artist_name: "Artist One",
            image_url: "/image1.jpg",
            album: {
              images: [{ url: "/image1.jpg" }],
            },
          },
          {
            id: "2",
            name: "Song Two",
            artist_name: "Artist Two",
            image_url: "/image2.jpg",
            album: {
              images: [{ url: "/image2.jpg" }],
            },
          },
        ],
      }),
  })
) as jest.Mock;

describe("DynamicDataDisplay Component", () => {
  const defaultProps = {
    endpoint: "/api/recommendations",
    type: "track",
    selectedSongs: [],
    onSelectSong: jest.fn(),
    onClearSelection: jest.fn(),
  };

  beforeAll(() => {
    global.ResizeObserver = class {
      observe() {}
      unobserve() {}
      disconnect() {}
    };
  });

  it("displays loading animation while fetching data", () => {
    // we can wrap this with await act, but loading is complete by using it
    render(<DynamicDataDisplay {...defaultProps} />);
    expect(
      screen.getByText("Hold tight, we're crunching some numbers")
    ).toBeInTheDocument();
    expect(
      screen.getByText(
        "This may take up to 60 seconds, but it’ll be worth the wait!"
      )
    ).toBeInTheDocument();
  });

  it("fetches and renders CardGrid when data is loaded and type is 'track'", async () => {
    await act(async () => {
      render(<DynamicDataDisplay {...defaultProps} />);
    });

    await waitFor(() => {
      expect(screen.getByText("Song One")).toBeInTheDocument();
    });

    expect(screen.getByText("Song Two")).toBeInTheDocument();
  });

  it("renders RecommendCardGrid when type is 'recommendations'", async () => {
    const props = { ...defaultProps, type: "recommendations" };

    await act(async () => {
      render(<DynamicDataDisplay {...props} />);
    });

    await waitFor(() => {
      expect(screen.getByText("Artist One")).toBeInTheDocument();
    });
  });

  it("handles error state", async () => {
    (global.fetch as jest.Mock).mockImplementationOnce(() =>
      Promise.resolve({
        json: () => Promise.resolve({ error: "Failed to fetch data" }),
      })
    );

    await act(async () => {
      render(<DynamicDataDisplay {...defaultProps} />);
    });

    await waitFor(() => {
      expect(screen.getByText("Failed to fetch data")).toBeInTheDocument();
    });
  });

  it("handles item click and selection", async () => {
    const mockOnSelectSong = jest.fn();

    await act(async () => {
      render(
        <DynamicDataDisplay
          {...defaultProps}
          onSelectSong={mockOnSelectSong}
          selectedSongs={[]}
        />
      );
    });

    await waitFor(() => {
      expect(screen.getByText("Song One")).toBeInTheDocument();
    });

    // Simulate clicking on a song
    const songOne = screen.getByText("Song One").closest("div");
    fireEvent.click(songOne!);

    expect(mockOnSelectSong).toHaveBeenCalled();
  });

  it("displays 'Clear Selection' button when there are selected songs", async () => {
    const selectedSongs = [
      {
        id: "1",
        name: "Song One",
        artist_name: "Artist One",
        image_url: "/image1.jpg",
      },
    ];

    await act(async () => {
      render(
        <DynamicDataDisplay {...defaultProps} selectedSongs={selectedSongs} />
      );
    });

    expect(screen.getByText("Clear Selection")).toBeInTheDocument();

    // Simulate clearing selection
    const clearButton = screen.getByText("Clear Selection");
    fireEvent.click(clearButton);

    expect(defaultProps.onClearSelection).toHaveBeenCalled();
  });

  it("renders correctly when no data is returned", async () => {
    (global.fetch as jest.Mock).mockImplementationOnce(() =>
      Promise.resolve({
        json: () => Promise.resolve({ items: [] }),
      })
    );

    await act(async () => {
      render(<DynamicDataDisplay {...defaultProps} />);
    });

    await waitFor(() => {
      expect(screen.queryByText("Song One")).not.toBeInTheDocument();
      expect(screen.queryByText("Song Two")).not.toBeInTheDocument();
    });
  });
});
