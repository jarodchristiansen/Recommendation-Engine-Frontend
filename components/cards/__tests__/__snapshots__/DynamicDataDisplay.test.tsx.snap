// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicDataDisplay Component matches the snapshot of the DynamicDataDisplay 1`] = `
<div>
  <div
    class="mt-4 p-6 bg-white shadow-md rounded-lg"
  >
    <div
      class="flex flex-col items-center text-center text-gray-500"
    >
      <svg
        class="animate-bounce text-5xl text-blue-500 mb-2"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 640 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M32,224H64V416H32A31.96166,31.96166,0,0,1,0,384V256A31.96166,31.96166,0,0,1,32,224Zm512-48V448a64.06328,64.06328,0,0,1-64,64H160a64.06328,64.06328,0,0,1-64-64V176a79.974,79.974,0,0,1,80-80H288V32a32,32,0,0,1,64,0V96H464A79.974,79.974,0,0,1,544,176ZM264,256a40,40,0,1,0-40,40A39.997,39.997,0,0,0,264,256Zm-8,128H192v32h64Zm96,0H288v32h64ZM456,256a40,40,0,1,0-40,40A39.997,39.997,0,0,0,456,256Zm-8,128H384v32h64ZM640,256V384a31.96166,31.96166,0,0,1-32,32H576V224h32A31.96166,31.96166,0,0,1,640,256Z"
        />
      </svg>
      <span
        class="text-xl font-semibold mb-2"
      >
        Hold tight, we're crunching some numbers
      </span>
      <p
        class="text-sm"
      >
        This may take up to 60 seconds, but itâ€™ll be worth the wait!
      </p>
    </div>
  </div>
</div>
`;
